src/poetry/core/factory.py

```diff
@@ -366,6 +367,10 @@ def create_dependency(
                     )
             else:
                 marker = parse_marker(markers)
+                # possibility that python constraint appears in markers, not keywords
+                if not python_versions and markers:
```

Comment:
`markers` is known truth-y here, isn't it, making this check redundant?

Having said which I wonder whether it would be ever so slightly clearer to do this up at line 260-ish, immediately after
```python
            python_versions = constraint.get("python")
            platform = constraint.get("platform")
            markers = constraint.get("markers")
```
(then the check would not be redundant)